pjnath/src/pjnath/turn_sock.c

```diff
@@ -1766,3 +1785,196 @@ static void turn_on_connection_bind_status(pj_turn_session *sess,
 					      peer_addr, addr_len);
     }
 }
+
+static void turn_on_connect(pj_turn_session *sess,
+                    pj_status_t status,
+                    pj_uint32_t conn_id,
+                    const pj_sockaddr_t *peer_addr,
+                    unsigned addr_len)
+{
+    pj_turn_sock *turn_sock = (pj_turn_sock*) 
+			      pj_turn_session_get_user_data(sess);
+    pj_pool_t *pool;
+    tcp_data_conn_t *new_conn;
+    pj_turn_session_info info;
+    pj_sock_t sock = PJ_INVALID_SOCKET;
+    pj_activesock_cfg asock_cfg;
+    pj_activesock_cb asock_cb;
+    pj_sockaddr bound_addr, *cfg_bind_addr;
+    pj_uint16_t max_bind_retry;
+    char addrtxt[PJ_INET6_ADDRSTRLEN+8];
+    unsigned i;
+
+    PJ_ASSERT_ON_FAIL(turn_sock->conn_type == PJ_TURN_TP_TCP &&
+		      turn_sock->alloc_param.peer_conn_type == PJ_TURN_TP_TCP,
+		      return);
+    PJ_LOG(5,(turn_sock->pool->obj_name, "Trying to connect to peer %s",
+	      pj_sockaddr_print(peer_addr, addrtxt, sizeof(addrtxt), 3)));
+
+    if (turn_sock == NULL) {
+	/* We've been destroyed */
+	return;
+    }
+
+    pj_grp_lock_acquire(turn_sock->grp_lock);
+
```

Comment:
Around here we should check & notify app if the connection attempt fails (i.e: `status != PJ_SUCCESS`) via `on_connection_status`.